java/arcs/core/storage/WriteBack.kt

```diff
@@ -0,0 +1,164 @@
+        if (!passThrough.value && scope != null) {
+            channel.consumeAsFlow()
+                .onEach {
+                    // Neither black out pending flush jobs in this channel nor propagate
+                    // the exception within the scope to affect flush jobs at other stores.
+                    exitFlushSection { try { it() } catch (_: Exception) {} }
+                }
+                .onCompletion {
+                    // Upon cancellation of the write-back scope, change to write-through mode,
+                    // consume all pending flush jobs then release all awaitings.
+                    passThrough.update { true }
+                    if (!channel.isEmpty && !channel.isClosedForReceive) {
+                        channel.consumeEach {
+                            exitFlushSection { try { it() } catch (_: Exception) {} }
+                        }
+                    }
+                    if (awaitSignal.isLocked) awaitSignal.unlock()
+                }
+                .launchIn(scope)
+        }
+    }
+
+    override suspend fun flush(job: suspend () -> Unit) {
+        if (!passThrough.value) flushSection { job() }
+        else job()
+    }
+
+    override suspend fun asyncFlush(job: suspend () -> Unit) {
+        if (!passThrough.value) enterFlushSection {
+            // Queue up a flush task can run average 3x-5x faster than launching it.
+            // Fall back to write-through when the write-back channel is cancelled or closed.
+            try { channel.send(job) } catch (_: Exception) { exitFlushSection { job() } }
+        } else job()
+    }
+
+    override suspend fun awaitIdle() = awaitSignal.withLock {}
+
+    private suspend inline fun flushSection(job: () -> Unit) {
+        enterFlushSection()
+        job()
+        exitFlushSection()
+    }
+
+    private suspend inline fun enterFlushSection(job: () -> Unit = {}) {
+        withLock { if (++activeJobs == 1) awaitSignal.lock() }
+        job()
+    }
+
+    private suspend inline fun exitFlushSection(job: () -> Unit = {}) {
+        job()
+        withLock { if (--activeJobs == 0) awaitSignal.unlock() }
+    }
+
+    companion object : WriteBackFactory {
+        private var writeBackScope: CoroutineScope? = null
+
+        /** Override [WriteBack] injection to Arcs Stores. */
+        var writeBackFactoryOverride: WriteBackFactory? = null
+
+        /** The factory of creating [WriteBack] instances. */
+        override fun create(protocol: String, queueSize: Int): WriteBack =
+            writeBackFactoryOverride?.create(protocol, queueSize)
+                ?: StoreWriteBack(protocol, queueSize, writeBackScope)
+
+        /**
+         * Initialize write-back coroutine scope.
+         * The caller is responsible for managing lifecycle of the [scope].
+         * The cancellation of the [scope] will switch the write-back to write-through.
+         */
+        fun init(scope: CoroutineScope) { writeBackScope = scope }
```

Comment:
Should this throw an exception if writeBackScope is non-null already?

Reply:
Add non-null check would fail several tests built on top of TestBindingDelegate which spins up StorageService instances per tests.

Multiple StorageService instances are impossible in the real world.
So I left it as it is :-)